### ISSUE TITLE
superblock journal seq / journal seq mismatch on multi-device FS

### ISSUE THREAD

**Issue by joelarmstrong** [koverstreet/bcachefs#7]
Hi,

Twice now I've run into what looks like a data corruption bug, one that fsck/`-o fix_errors` seems not to be able to fix easily. (I make daily backups, so nothing was lost.) The rough sequence of steps I followed each time:

1. Set up an encrypted multi-device tiered filesystem, /dev/sda2 (HDD) as background_target, /dev/sdb2 (SSD) as foreground_target and promote_target. This serves as my root filesystem which has to be mounted using a hacked up initramfs.
2. At some later point set up background_compression with zstd.
3. Everything works fine over multiple restarts, until:
4. At some point there is an attempt by the initramfs to mount *one* of the drives, not both, read-only. For me this was caused by a) accidentally running an old initramfs or b) my hacked up initramfs being blown away by an upgrade. Anyway, this fails with `insufficient devices`, as expected, but takes a noticeably long time to do so.
```
[   43.728424] bcachefs: bch2_fs_open() bch_fs_open err opening /dev/sda2: insufficient devices
```
5. When trying to mount the filesystem after that (properly, with both devices), I get this error:
```
[   81.943979] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): recovering from clean shutdown, journal seq 18446744073709551615
[   86.232951] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): journal read done, 3781 keys in 113 entries, seq 331620
[   86.232954] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): superblock journal seq (18446744073709551615) doesn't match journal (331620) a
[   86.233018] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): Unable to continue, halting
[   86.233085] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): filesystem contains errors: please report this to the developers
[   86.233145] mount with -o fix_errors to repair
[   86.233146] bcachefs: bch2_fs_open() bch_fs_open err opening /dev/sdb2: fsck error
```
NB: the superblock journal seq is 2^64 - 1.

6. The FS is difficult to recover at this point; `-o fix_errors` doesn't help, and fsck seems to get a little farther, but mount then fails again. Multiple rounds of fsck/mount/fsck/mount don't help.

I can't be totally sure what causes this. Step 4 seems necessary to trigger this bug, but not sufficient: it directly preceded the error both times. But I tried replicating this with smaller file-backed loopback devices without any real success, so maybe there needs to be some specific data in the journal for it to trigger.

This is all with commit 9c08efd02b827e6ee2a6e5a751045c6c6bd484d4, so a couple weeks out of date now. I'll upgrade and see if it triggers again.

Here's the superblock information (taken after the mount failed) and a dmesg log if it helps. Note that the "clean" flag is different on the two devices, not sure if that matters.

/dev/sda2 superblock:
```
External UUID:                  1c6fb4a7-6172-4646-9dda-64da58327beb
Internal UUID:                  a07f676a-de61-43a7-a1b3-a18d8f813344
Label:
Version:                        9
Block_size:                     512
Btree node size:                256.0K
Error action:                   remount-ro
Clean:                          1
Metadata replicas:              1
Data replicas:                  1
Metadata checksum type:         crc32c (1)
Data checksum type:             crc32c (1)
Compression type:               none (0)
Foreground write target:        1
Background write target:        2
Promote target:                 1
String hash type:               siphash (2)
32 bit inodes:                  0
GC reserve percentage:          8%
Root reserve percentage:        0%
Devices:                        2 live, 2 total
Sections:                       journal,members,crypt,replicas,clean
Superblock size:                9200
Members (size 120):
  Device 0:
    UUID:                       7b799051-7060-4e77-9144-72b5f184b171
    Size:                       223.1G
    Bucket size:                256.0K
    First bucket:               0
    Buckets:                    913696
    Last mount:                 Thu Jan  1 00:00:46 1970
    State:                      readwrite
    Group:                      `�
    Data allowed:               journal,btree,data
    Has data:                   journal,btree,data,cached
    Replacement policy:         lru
    Discard:                    0
  Device 1:
    UUID:                       19f52b7a-ba13-4b4c-8f71-de883e541c8c
    Size:                       1.8T
    Bucket size:                512.0K
    First bucket:               0
    Buckets:                    3815444
    Last mount:                 Thu Jan  1 00:00:46 1970
    State:                      readwrite
    Group:                      `�
    Data allowed:               journal,btree,data
    Has data:                   journal,btree,data
    Replacement policy:         lru
    Discard:                    0
```

/dev/sdb2 superblock:
```
External UUID:                  1c6fb4a7-6172-4646-9dda-64da58327beb
Internal UUID:                  a07f676a-de61-43a7-a1b3-a18d8f813344
Label:
Version:                        9
Block_size:                     512
Btree node size:                256.0K
Error action:                   remount-ro
Clean:                          0
Metadata replicas:              1
Data replicas:                  1
Metadata checksum type:         crc32c (1)
Data checksum type:             crc32c (1)
Compression type:               none (0)
Foreground write target:        1
Background write target:        2
Promote target:                 1
String hash type:               siphash (2)
32 bit inodes:                  0
GC reserve percentage:          8%
Root reserve percentage:        0%
Devices:                        2 live, 2 total
Sections:                       journal,members,crypt,replicas,clean
Superblock size:                9480
Members (size 120):
  Device 0:
    UUID:                       7b799051-7060-4e77-9144-72b5f184b171
    Size:                       223.1G
    Bucket size:                256.0K
    First bucket:               0
    Buckets:                    913696
    Last mount:                 Thu Jan  1 00:00:46 1970
    State:                      readwrite
    Group:                      �
    Data allowed:               journal,btree,data
    Has data:                   journal,btree,data,cached
    Replacement policy:         lru
    Discard:                    0
  Device 1:
    UUID:                       19f52b7a-ba13-4b4c-8f71-de883e541c8c
    Size:                       1.8T
    Bucket size:                512.0K
    First bucket:               0
    Buckets:                    3815444
    Last mount:                 Thu Jan  1 00:00:46 1970
    State:                      readwrite
    Group:                      �
    Data allowed:               journal,btree,data
    Has data:                   journal,btree,data
    Replacement policy:         lru
    Discard:                    0
```

diff between the two, for clarity:
```
8c8
< Clean:                                1
---
> Clean:                                0
23c23
< Superblock size:              9200
---
> Superblock size:              9480
35c35
<     Group:                    `�
---
>     Group:                    �
49c49
<     Group:                    `�
---
>     Group:                    �
```

dmesg log with `verbose_init,verbose_recovery` options:
```
[ 2357.322991] bcachefs: bch2_fs_open()
[ 2357.322992] bcachefs: bch2_read_super()
[ 2357.323669] bcachefs: bch2_read_super() ret 0
[ 2357.324021] bcachefs: bch2_read_super()
[ 2357.326757] bcachefs: bch2_read_super() ret 0
[ 2357.327053] bcachefs: bch2_fs_alloc()
[ 2357.327354] bcachefs: bch2_fs_journal_init()
[ 2357.327359] bcachefs: bch2_fs_journal_init() ret 0
[ 2357.327360] bcachefs: bch2_fs_btree_cache_init()
[ 2357.327427] bcachefs: bch2_fs_btree_cache_init() ret 0
[ 2357.327462] bcachefs: bch2_fs_encryption_init()
[ 2357.327475] bcachefs: bch2_fs_encryption_init() ret 0
[ 2357.327477] bcachefs: __bch2_fs_compress_init()
[ 2357.327479] bcachefs: __bch2_fs_compress_init() ret 0
[ 2357.327480] bcachefs: bch2_fs_fsio_init()
[ 2357.327492] bcachefs: bch2_fs_fsio_init() ret 0
[ 2357.327492] bcachefs: bch2_dev_alloc()
[ 2357.328782] bcachefs: bch2_dev_alloc() ret 0
[ 2357.328784] bcachefs: bch2_dev_alloc()
[ 2357.333838] bcachefs: bch2_dev_alloc() ret 0
[ 2357.334022] bcachefs: bch2_fs_alloc() ret 0
[ 2357.334059] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): recovering from clean shutdown, journal seq 18446744073709551615
[ 2361.643306] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): journal read done, 3781 keys in 113 entries, seq 331620
[ 2361.643308] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): superblock journal seq (18446744073709551615) doesn't match journal (331620) after clean shutdown, exiting
[ 2361.643378] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): Unable to continue, halting
[ 2361.643444] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): filesystem contains errors: please report this to the developers
[ 2361.643505] mount with -o fix_errors to repair
[ 2361.643506] bcachefs: bch2_fs_open() bch_fs_open err opening /dev/sda2: fsck error
[ 2361.643564] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): shutting down
[ 2361.689327] bcachefs (1c6fb4a7-6172-4646-9dda-64da58327beb): shutdown complete
[ 2361.704344] bcachefs: bch2_fs_open() ret -22
```

**Comment by koverstreet** [comment]
finally fixed by commit 00c04f8485

**Comment by joelarmstrong** [comment]
Thanks for tracking this down!
